clients/web/src/views/App.js

```diff
@@ -31,6 +31,19 @@ import 'girder/stylesheets/layout/global.styl';
 import 'girder/stylesheets/layout/layout.styl';
 
 var App = View.extend({
+    events: {
+        'click a': (evt) => {
+            if (!evt.isDefaultPrevented()) {
+                var elem = $(evt.target),
+                    href = elem.attr('href');
+                if (elem.is('a') && href && href.substr(0, 1) === '#') {
+                    router.navigate(href.substr(1), {trigger: true});
+                    evt.preventDefault();
```

Comment:
Does this code make other people uncomfortable, or is it just me?

During a meeting someone (perhaps jokingly) insinuated that part of the reason this problem was so hellish was because of either Backbone or Girder's idiosyncrasies. Since we're seriously entertaining the idea of a rewrite in a new framework, I think it would be a great exercise to try and figure out how this problem would be addressed in a different candidate framework (Vue or React).

Reply:
We could get rid of this code; it allows `$(<a element with href>).click()` to work, which prevents breaking tests AND is nice when poking around manually.  It could also be moved to just the testing app, rather than in the real app (but any differences between the testing app and the real app could hide bugs because the behavior is different).

If we had used hrefs as the preferred method of routing wherever possible, this wouldn't have been neceesary -- all of our tests would have used `$(<a element with href>)[0].click()` if they were using jQuery (and that now works, too).

We have a tension between wanting to make things convenient for the general case, wanting code purity, wanting to make testing easy, wanting to be able to customize everything, wanting to have API stability, and wanting many features.  I proposed this change because I think it gets us a decent balance of these goals; there are other ways to do it -- we could break existing tests, we could make the test app and real app different, we could refuse to have the utility of hrefs, etc.  There may be a better way to accomplish the same thing (or a better balance of the goals).

If we do a major rewrite, I think we should use hrefs as routing wherever possible; it surprises the user less and builds in functionality of being able to restore state.

Reply:
Thanks for the explanation. If this is mostly just to get our links to work as expected in test code, I would be more comfortable putting it just in testing code, though the risk of hiding bugs as a result is noted.

> If we do a major rewrite, I think we should use hrefs as routing wherever possible

Definitely agreed, I think all of us are on board with that. I'm particularly interested in how other frameworks handle the pattern of widgets that sometimes need to do routing and other times need to emit events when elements are clicked.

One bit of good news is that a rewrite would probably allow us to jettison jquery altogether, which would make this testing headache go away.

Reply:
I briefly thought about this myself. I think new View framework will be able to hide complex logic inside a component so it's clean to use and looks clean like [this](https://reacttraining.com/react-router/web/api/Link), but internally it's doing something like [this](https://github.com/STRML/react-router-component/commit/9984d765d01d4e8e67d3da7eb39b035253e62925).

Reply:
Indeed, both `react-router` and `vue-router`, do the "right thing", in that even when you use their native syntax for adding rotatable elements, they end up creating `<a href...`> elements in the DOM/